### PR TITLE
bpo-36854: Move GC runtime state from _PyRuntimeState to PyInterpreterState.

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
Note: as part of this PR, I've made sure that Python threads (created by the threading module) are marked as "deleted" earlier in finalization.  That marker is how joining threads get unblocked.  It was happening in `PyThreadState_Delete()` but now it will happen in `PyThreadState_Clear()`.  This is necessary to ensure that the callback gets called *before* much interpreter/runtime finalization happens.  (It could impact daemon threads, but we already can't guarantee behavior for those once finalization starts.) 

https://bugs.python.org/issue36854

### CODE REVIEW

File: Python/pystate.c [hunk 1]
```diff
@@ -782,6 +781,13 @@ PyThreadState_Clear(PyThreadState *tstate)
     Py_CLEAR(tstate->async_gen_finalizer);
 
     Py_CLEAR(tstate->context);
+
+    if (tstate->on_delete != NULL) {
+        // This will unblock any joining threads.
+        tstate->on_delete(tstate->on_delete_data);
+        tstate->on_delete = NULL;
+        tstate->on_delete_data = NULL;
```
Comment (vstinner) [inline Python/pystate.c:789]: This change looks to be unrelated. If it's related, I would prefer to see it as a separated PR to prepare that one.

Same comment for other thread changes:

```
    // All threading module threads are marked as "done" later
    // in PyThreadState_Clear().
```

and

```
        // This will unblock any joining threads.
        // We also do this in PyThreadState_Clear(), but do it here to be sure.
```

File: Include/internal/pycore_object.h [hunk 2]
```diff
@@ -38,11 +38,12 @@ static inline void _PyObject_GC_TRACK_impl(const char *filename, int lineno,
                           "object is in generation which is garbage collected",
                           filename, lineno, "_PyObject_GC_TRACK");
 
-    PyGC_Head *last = (PyGC_Head*)(_PyRuntime.gc.generation0->_gc_prev);
+    PyInterpreterState *interp = _PyInterpreterState_GET_UNSAFE();
```
Comment (vstinner) [inline Include/internal/pycore_object.h:41]: Would it be possible to use a "struct _gc_runtime_state *state" variable instead?

```
struct _gc_runtime_state *state = &_PyInterpreterState_GET_UNSAFE()->gc;
```

File: Modules/gcmodule.c [hunk 3]
```diff
@@ -1257,7 +1257,8 @@ static PyObject *
 gc_enable_impl(PyObject *module)
 /*[clinic end generated code: output=45a427e9dce9155c input=81ac4940ca579707]*/
 {
-    _PyRuntime.gc.enabled = 1;
+    PyInterpreterState *interp = _PyInterpreterState_GET_UNSAFE();
```
Comment (vstinner) [inline Modules/gcmodule.c:1260]: Would it be possible to use a "struct _gc_runtime_state *state" variable instead?

```
struct _gc_runtime_state *state = &_PyInterpreterState_GET_UNSAFE()->gc;
```

If the line is too lone, maybe add a GET_STATE() macro which returns ``&_PyInterpreterState_GET_UNSAFE()->gc``.

Same comment for the whole file. I would prefer to avoid introducing "PyInterpreterState *interp = _PyInterpreterState_GET_UNSAFE(); " lines if it's only used to get the GC state.

File: Objects/object.c [hunk 4]
```diff
@@ -2074,11 +2075,12 @@ _PyTrash_thread_deposit_object(PyObject *op)
 void
 _PyTrash_destroy_chain(void)
 {
-    while (_PyRuntime.gc.trash_delete_later) {
-        PyObject *op = _PyRuntime.gc.trash_delete_later;
+    PyInterpreterState *interp = _PyInterpreterState_GET_UNSAFE();
```
Comment (vstinner) [inline Objects/object.c:2078]: Again, a "state" variable would be more appropriate here.

File: Python/pylifecycle.c [hunk 5]
```diff
@@ -511,6 +511,8 @@ pycore_create_interpreter(_PyRuntimeState *runtime,
     }
     *interp_p = interp;
 
+    _PyGC_Initialize(&interp->gc);
```
Comment (vstinner) [inline Python/pylifecycle.c:514]: Hum, maybe _PyGC_Initialize() should take interp argument instead, to be consistent with other _PyGC functions.

**Review by vstinner — CHANGES_REQUESTED** [review]
I like what I see: -) But I have some comments on the actual implementation.

**Comment by csabella** [comment]
@ericsnowcurrently, were you wanting to land this for 3.8?

**Comment by vstinner** [comment]
I wrote a different change, PR #17287, that I just merged. So I close this old PR. Thanks anyway Eric for working on this, my work is partially based on yours ;-)

**Comment by ericsnowcurrently** [comment]
Thanks for doing it!!!

**Comment by vstinner** [comment]
> Thanks for doing it!!!

You're welcome.
